File: app/src/main/java/com/salman/bitclock/data/database/AlarmDao.java
package com.salman.bitclock.data.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.salman.bitclock.data.models.Alarm;

import java.util.List;

@Dao
public interface AlarmDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(Alarm alarm);

    @Update
    void update(Alarm alarm);

    @Delete
    void delete(Alarm alarm);

    @Query("SELECT * FROM alarms ORDER BY hour, minute ASC")
    LiveData<List<Alarm>> getAllAlarms();
    
    @Query("SELECT * FROM alarms WHERE id = :alarmId")
    LiveData<Alarm> getAlarmById(int alarmId);

    @Query("SELECT * FROM alarms")
    List<Alarm> getAllAlarmsNonLiveData();

    // **FIX**: Added for synchronous fetch in AlarmReceiver
    @Query("SELECT * FROM alarms WHERE id = :alarmId")
    Alarm getAlarmByIdSync(int alarmId);
}

File: app/src/main/kotlin/com/salman/bitclock/data/database/AlarmDao.kt
package com.salman.bitclock.data.database

import androidx.room.*
import com.salman.bitclock.data.models.Alarm
import kotlinx.coroutines.flow.Flow

@Dao
interface AlarmDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(alarm: Alarm): Long

    @Update
    suspend fun update(alarm: Alarm)

    @Delete
    suspend fun delete(alarm: Alarm)

    @Query("SELECT * FROM alarms ORDER BY hour, minute ASC")
    fun getAllAlarms(): Flow<List<Alarm>>
    
    @Query("SELECT * FROM alarms WHERE id = :alarmId")
    fun getAlarmById(alarmId: Int): Flow<Alarm?>

    @Query("SELECT * FROM alarms")
    suspend fun getAllAlarmsSync(): List<Alarm>

    @Query("SELECT * FROM alarms WHERE id = :alarmId")
    suspend fun getAlarmByIdSync(alarmId: Int): Alarm?
}
